### ISSUE TITLE
Add real examples for each design pattern

### ISSUE THREAD

**Issue by Torpedr** [domnikl/DesignPatternsPHP#268]
Hi, I have a more general suggestion to add a real example problem that every design pattern solves.

The examples only tell us where this particular pattern has been used but don't actually help a beginner to decide _when_ to use them.

For instance, strategy pattern may solve the following problems:
 - sorting,
 - input data validation,
 - types of file encoding,
 and so on.

Thanks :)

**Comment by ihor** [comment]
I'm not the maintainer of this repo but I'll say something you can hear from many experienced programmers.

Rule of thumb – if you are not sure whether to use a pattern, then don't :) Patterns won't make your code better. They are just ideas how you can solve some common problems.

The main problem is that patterns don't (in some cases just badly) explain why this solution is better than other solutions to this particular problem. As result, they don't teach you how to change your way of thinking to become a better software engineer.

The best thing beginners could do is to stop thinking in patterns and switch to principles. Which come from values. Which depend on the problem you are trying to solve.

To summarize, real life examples of patterns without tons of explanations won't help to understand why that particular solution was used. These examples won't be better than the synthetic examples you can find here. Look at this repo like at the catalog of possible solutions to common problems. Nothing more.

**Comment by domnikl** [comment]
@Torpedr like @ihor said: it won't help to list real-world examples here, because each problem is unique and there is no common problem -> pattern mapping. There are always multiple solutions to a problem. 

Unfortunately, there is no fast path to software architecture for beginners.
